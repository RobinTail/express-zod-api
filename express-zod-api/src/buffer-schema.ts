import { z } from "zod/v4";

export const bufferSchema = z.custom<Buffer>(
  (subject) => Buffer.isBuffer(subject),
  { error: "Expected Buffer" },
);
