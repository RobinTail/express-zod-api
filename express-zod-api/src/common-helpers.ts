import { Request } from "express";
import * as R from "ramda";
import { z } from "zod";
import { CommonConfig, InputSource, InputSources } from "./config-type";
import { contentTypes } from "./content-type";
import { OutputValidationError } from "./errors";
import { metaSymbol } from "./metadata";
import { AuxMethod, Method } from "./method";

/** @desc this type does not allow props assignment, but it works for reading them when merged with another interface */
export type EmptyObject = Record<string, never>;
export type EmptySchema = z.ZodObject<EmptyObject, "strip">;
export type FlatObject = Record<string, unknown>;

/** @link https://stackoverflow.com/a/65492934 */
type NoNever<T, F> = [T] extends [never] ? F : T;

/**
 * @desc Using module augmentation approach you can specify tags as the keys of this interface
 * @example declare module "express-zod-api" { interface TagOverrides { users: unknown } }
 * @link https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation
 * */
// eslint-disable-next-line @typescript-eslint/no-empty-object-type -- augmentation
export interface TagOverrides {}
export type Tag = NoNever<keyof TagOverrides, string>;

/** @see https://expressjs.com/en/guide/routing.html */
export const routePathParamsRegex = /:([A-Za-z0-9_]+)/g;
export const getRoutePathParams = (path: string): string[] =>
  path.match(routePathParamsRegex)?.map((param) => param.slice(1)) || [];

const areFilesAvailable = (request: Request): boolean => {
  const contentType = request.header("content-type") || "";
  const isUpload = contentType.toLowerCase().startsWith(contentTypes.upload);
  return "files" in request && isUpload;
};

export const defaultInputSources: InputSources = {
  get: ["query", "params"],
  post: ["body", "params", "files"],
  put: ["body", "params"],
  patch: ["body", "params"],
  delete: ["query", "params"],
};
const fallbackInputSource: InputSource[] = ["body", "query", "params"];

export const getActualMethod = (request: Request) =>
  request.method.toLowerCase() as Method | AuxMethod;

export const getInput = (
  req: Request,
  userDefined: CommonConfig["inputSources"] = {},
): FlatObject => {
  const method = getActualMethod(req);
  if (method === "options") return {};
  return (
    userDefined[method] ||
    defaultInputSources[method] ||
    fallbackInputSource
  )
    .filter((src) => (src === "files" ? areFilesAvailable(req) : true))
    .reduce<FlatObject>((agg, src) => Object.assign(agg, req[src]), {});
};

export const ensureError = (subject: unknown): Error =>
  subject instanceof Error ? subject : new Error(String(subject));

export const getMessageFromError = (error: Error): string => {
  if (error instanceof z.ZodError) {
    return error.issues
      .map(({ path, message }) =>
        (path.length ? [path.join("/")] : []).concat(message).join(": "),
      )
      .join("; ");
  }
  if (error instanceof OutputValidationError) {
    const hasFirstField = error.cause.issues[0]?.path.length > 0;
    return `output${hasFirstField ? "/" : ": "}${error.message}`;
  }
  return error.message;
};

/** Takes the original unvalidated examples from the properties of ZodObject schema shape */
export const pullExampleProps = <T extends z.SomeZodObject>(subject: T) =>
  Object.entries(subject.shape).reduce<Partial<z.input<T>>[]>(
    (acc, [key, schema]) => {
      const { _def } = schema as z.ZodType;
      return combinations(
        acc,
        (_def[metaSymbol]?.examples || []).map(R.objOf(key)),
        ([left, right]) => ({ ...left, ...right }),
      );
    },
    [],
  );

export const getExamples = <
  T extends z.ZodType,
  V extends "original" | "parsed" | undefined,
>({
  schema,
  variant = "original",
  validate = variant === "parsed",
  pullProps = false,
}: {
  schema: T;
  /**
   * @desc examples variant: original or parsed
   * @example "parsed" — for the case when possible schema transformations should be applied
   * @default "original"
   * @override validate: variant "parsed" activates validation as well
   * */
  variant?: V;
  /**
   * @desc filters out the examples that do not match the schema
   * @default variant === "parsed"
   * */
  validate?: boolean;
  /**
   * @desc should pull examples from properties — applicable to ZodObject only
   * @default false
   * */
  pullProps?: boolean;
}): ReadonlyArray<V extends "parsed" ? z.output<T> : z.input<T>> => {
  let examples = schema._def[metaSymbol]?.examples || [];
  if (!examples.length && pullProps && schema instanceof z.ZodObject)
    examples = pullExampleProps(schema);
  if (!validate && variant === "original") return examples;
  const result: Array<z.input<T> | z.output<T>> = [];
  for (const example of examples) {
    const parsedExample = schema.safeParse(example);
    if (parsedExample.success)
      result.push(variant === "parsed" ? parsedExample.data : example);
  }
  return result;
};

export const combinations = <T>(
  a: T[],
  b: T[],
  merge: (pair: [T, T]) => T,
): T[] => (a.length && b.length ? R.xprod(a, b).map(merge) : a.concat(b));

/**
 * @desc isNullable() and isOptional() validate the schema's input
 * @desc They always return true in case of coercion, which should be taken into account when depicting response
 */
export const hasCoercion = (schema: z.ZodTypeAny): boolean =>
  "coerce" in schema._def && typeof schema._def.coerce === "boolean"
    ? schema._def.coerce
    : false;

export const ucFirst = (subject: string) =>
  subject.charAt(0).toUpperCase() + subject.slice(1).toLowerCase();

export const makeCleanId = (...args: string[]) => {
  const byAlpha = R.chain((entry) => entry.split(/[^A-Z0-9]/gi), args);
  const byWord = R.chain(
    (entry) =>
      entry.replaceAll(/[A-Z]+/g, (beginning) => `/${beginning}`).split("/"),
    byAlpha,
  );
  return byWord.map(ucFirst).join("");
};

export const getTransformedType = R.tryCatch(
  <T>(schema: z.ZodEffects<z.ZodTypeAny, unknown, T>, sample: T) =>
    typeof schema.parse(sample),
  R.always(undefined),
);

/** @desc can still be an array, use Array.isArray() or rather R.type() to exclude that case */
export const isObject = (subject: unknown) =>
  typeof subject === "object" && subject !== null;

export const isProduction = R.memoizeWith(
  () => process.env.TSUP_STATIC as string, // eslint-disable-line no-restricted-syntax -- substituted by TSUP
  () => process.env.NODE_ENV === "production", // eslint-disable-line no-restricted-syntax -- memoized
);
