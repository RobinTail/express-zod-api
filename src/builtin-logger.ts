import { Ansis, blue, cyanBright, green, hex, red } from "ansis";
import { inspect } from "node:util";
import { performance } from "node:perf_hooks";
import type { FlatObject } from "./common-helpers";
import { AbstractLogger, formatDuration, severity } from "./logger-helpers";

interface Context extends FlatObject {
  requestId?: string;
}

export interface BuiltinLoggerConfig {
  /**
   * @desc The minimal severity to log or "silent" to disable logging
   * @example "debug" also enables pretty output for inspected entities
   * */
  level: "silent" | "warn" | "info" | "debug";
  /**
   * @desc Enables colors on printed severity and inspected entities
   * @default Ansis::isSupported()
   * */
  color?: boolean;
  /**
   * @desc Control how deeply entities should be inspected
   * @default 2
   * @example null
   * @example Infinity
   * */
  depth?: number | null;
  /**
   * @desc Context: the metadata applicable for each logged entry, used by .child() method
   * @see childLoggerProvider
   * */
  ctx?: Context;
}

interface ProfilerOptions {
  message: string;
  /** @default "debug" */
  severity?: keyof AbstractLogger | ((ms: number) => keyof AbstractLogger);
  /** @default formatDuration - adaptive units and limited fraction */
  formatter?: (ms: number) => string | number;
}

/** @desc Built-in console logger with optional colorful inspections */
export class BuiltinLogger implements AbstractLogger {
  protected hasColor: boolean;
  protected readonly styles: Record<keyof AbstractLogger, Ansis> = {
    debug: blue,
    info: green,
    warn: hex("#FFA500"),
    error: red,
  };

  /** @example new BuiltinLogger({ level: "debug", color: true, depth: 4 }) */
  public constructor(protected config: BuiltinLoggerConfig) {
    const { color: hasColor = new Ansis().isSupported() } = config;
    this.hasColor = hasColor;
  }

  protected prettyPrint(subject: unknown) {
    const { depth = 2 } = this.config;
    return inspect(subject, {
      depth,
      colors: this.hasColor,
      breakLength: this.config.level === "debug" ? 80 : Infinity,
      compact: this.config.level === "debug" ? 3 : true,
    });
  }

  protected print(
    method: keyof AbstractLogger,
    message: string,
    meta?: unknown,
  ) {
    if (
      this.config.level === "silent" ||
      severity[method] < severity[this.config.level]
    ) {
      return;
    }
    const { requestId, ...ctx } = this.config.ctx || {};
    const output: string[] = [new Date().toISOString()];
    if (requestId) {
      output.push(this.hasColor ? cyanBright(requestId) : requestId);
    }
    output.push(
      this.hasColor ? `${this.styles[method](method)}:` : `${method}:`,
      message,
    );
    if (meta !== undefined) {
      output.push(this.prettyPrint(meta));
    }
    if (Object.keys(ctx).length > 0) {
      output.push(this.prettyPrint(ctx));
    }
    console.log(output.join(" "));
  }

  public debug(message: string, meta?: unknown) {
    this.print("debug", message, meta);
  }

  public info(message: string, meta?: unknown) {
    this.print("info", message, meta);
  }

  public warn(message: string, meta?: unknown) {
    this.print("warn", message, meta);
  }

  public error(message: string, meta?: unknown) {
    this.print("error", message, meta);
  }

  public child(ctx: Context) {
    return new BuiltinLogger({ ...this.config, ctx });
  }

  /** @desc Measures the duration until you invoke the returned callback */
  public profile(message: string): () => void;
  public profile(options: ProfilerOptions): () => void;
  public profile(subject: string | ProfilerOptions) {
    const start = performance.now();
    return () => {
      const duration = performance.now() - start;
      const {
        message,
        severity = "debug",
        formatter = formatDuration,
      } = typeof subject === "object" ? subject : { message: subject };
      this.print(
        typeof severity === "function" ? severity(duration) : severity,
        message,
        formatter(duration),
      );
    };
  }
}
