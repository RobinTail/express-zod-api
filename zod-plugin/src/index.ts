import "./runtime.ts"; // side effects here
export { pack, unpack } from "./packer.ts";
export { getBrand } from "./brand.ts";
