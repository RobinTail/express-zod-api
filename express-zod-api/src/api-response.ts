import { z } from "zod";
import { isSchema } from "./common-helpers";
import { contentTypes } from "./content-type";

export const defaultStatusCodes = {
  positive: 200,
  negative: 400,
} satisfies Record<string, number>;

export type ResponseVariant = keyof typeof defaultStatusCodes;
export const responseVariants = Object.keys(
  defaultStatusCodes,
) as ResponseVariant[];

/** @public this is the user facing configuration */
export class ApiResponse<S extends z.ZodType> {
  readonly #schema: S;
  readonly #statusCodes?: [number, ...number[]];
  readonly #mimeTypes?: [string, ...string[]] | null;

  public constructor(
    subject:
      | S // shorthand for { schema: S }, default status codes and MIME types applied
      | {
          schema: S;
          /** @default 200 for a positive and 400 for a negative response */
          statusCode?: number | [number, ...number[]];
          /**
           * @example null is for no content, such as 204 and 302
           * @default "application/json"
           * */
          mimeType?: string | [string, ...string[]] | null;
        },
  ) {
    const { schema, statusCode, mimeType } = isSchema(subject)
      ? { schema: subject }
      : subject;
    this.#schema = schema;
    this.#statusCodes =
      typeof statusCode === "number" ? [statusCode] : statusCode;
    this.#mimeTypes = typeof mimeType === "string" ? [mimeType] : mimeType;
  }

  public normalize(fbStatus: number) {
    return {
      schema: this.#schema,
      statusCodes: this.#statusCodes ?? [fbStatus],
      mimeTypes:
        this.#mimeTypes === undefined ? [contentTypes.json] : this.#mimeTypes,
    };
  }
}

export type NormalizedResponse = ReturnType<
  ApiResponse<z.ZodType>["normalize"]
>;
