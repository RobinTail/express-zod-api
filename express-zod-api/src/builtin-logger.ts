import ansis from "ansis";
import { inspect } from "node:util";
import { performance } from "node:perf_hooks";
import { FlatObject, isProduction } from "./common-helpers";
import {
  AbstractLogger,
  formatDuration,
  isHidden,
  Severity,
  styles,
} from "./logger-helpers";

interface Context extends FlatObject {
  requestId?: string;
}

export interface BuiltinLoggerConfig {
  /**
   * @desc The minimal severity to log or "silent" to disable logging
   * @example "debug" also enables pretty output for inspected entities
   * */
  level: "silent" | "warn" | "info" | "debug";
  /** @desc Enables colors on printed severity and inspected entities */
  color: boolean;
  /**
   * @desc Printing logs does not delay execution
   * @see https://expressjs.com/en/advanced/best-practice-performance.html#do-logging-correctly
   * */
  async: boolean;
  /**
   * @desc Control how deeply entities should be inspected
   * @example null
   * @example Infinity
   * */
  depth: number | null;
  /**
   * @desc Context: the metadata applicable for each logged entry, used by .child() method
   * @see childLoggerProvider
   * */
  ctx: Context;
}

interface ProfilerOptions {
  message: string;
  /** @default "debug" */
  severity?: Severity | ((ms: number) => Severity);
  /** @default formatDuration - adaptive units and limited fraction */
  formatter?: (ms: number) => string | number;
}

/** @desc Built-in console logger with optional colorful inspections */
export class BuiltinLogger implements AbstractLogger {
  protected readonly config: BuiltinLoggerConfig;
  protected readonly stack: string[] = [];
  protected postponed?: NodeJS.Immediate;

  /** @example new BuiltinLogger({ level: "debug", color: true, depth: 4 }) */
  public constructor({
    color = ansis.isSupported(),
    level = isProduction() ? "warn" : "debug",
    async = isProduction(),
    depth = 2,
    ctx = {},
  }: Partial<BuiltinLoggerConfig> = {}) {
    this.config = { color, level, async, depth, ctx };
  }

  protected format(subject: unknown) {
    const { depth, color: colors, level } = this.config;
    return inspect(subject, {
      depth,
      colors,
      breakLength: level === "debug" ? 80 : Infinity,
      compact: level === "debug" ? 3 : true,
    });
  }

  protected print(method: Severity, message: string, meta?: unknown) {
    const {
      level,
      ctx: { requestId, ...ctx },
      color: hasColor,
      async: isAsync,
    } = this.config;
    if (level === "silent" || isHidden(method, level)) return;
    const output: string[] = [new Date().toISOString()];
    if (requestId) output.push(hasColor ? styles.ctx(requestId) : requestId);
    output.push(
      hasColor ? `${styles[method](method)}:` : `${method}:`,
      message,
    );
    if (meta !== undefined) output.push(this.format(meta));
    if (Object.keys(ctx).length > 0) output.push(this.format(ctx));
    (isAsync ? this.postpone.bind(this) : console.log)(output.join(" "));
  }

  protected postpone(line: string) {
    this.stack.push(line);
    clearImmediate(this.postponed);
    this.postponed = setImmediate(this.purge.bind(this)).unref(); // do not block process.exit()
  }

  /** @internal */
  public purge() {
    while (this.stack.length) console.log(this.stack.shift());
  }

  /** @internal */
  public dispose(cb: () => void) {
    this.config.level = "silent"; // no more logs
    clearImmediate(this.postponed);
    this.postponed = setImmediate(() => {
      this.purge();
      cb();
    });
  }

  public debug(message: string, meta?: unknown) {
    this.print("debug", message, meta);
  }

  public info(message: string, meta?: unknown) {
    this.print("info", message, meta);
  }

  public warn(message: string, meta?: unknown) {
    this.print("warn", message, meta);
  }

  public error(message: string, meta?: unknown) {
    this.print("error", message, meta);
  }

  public child(ctx: Context) {
    return new BuiltinLogger({ ...this.config, ctx });
  }

  /**
   * @desc The argument used for instance created by .child() method
   * @see ChildLoggerProvider
   * */
  public get ctx() {
    return this.config.ctx;
  }

  /** @desc Measures the duration until you invoke the returned callback */
  public profile(message: string): () => void;
  public profile(options: ProfilerOptions): () => void;
  public profile(subject: string | ProfilerOptions) {
    const start = performance.now();
    return () => {
      const duration = performance.now() - start;
      const {
        message,
        severity = "debug",
        formatter = formatDuration,
      } = typeof subject === "object" ? subject : { message: subject };
      this.print(
        typeof severity === "function" ? severity(duration) : severity,
        message,
        formatter(duration),
      );
    };
  }
}
