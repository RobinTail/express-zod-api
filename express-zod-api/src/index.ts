import "@express-zod-api/zod-plugin"; // side effects here
export { createConfig } from "./config-type.ts";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory.ts";
export { getMessageFromError } from "./common-helpers.ts";
export { ensureHttpError } from "./result-helpers.ts";
export { BuiltinLogger } from "./builtin-logger.ts";
export { Middleware } from "./middleware.ts";
export {
  ResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler.ts";
export { DependsOnMethod } from "./depends-on-method.ts";
export { ServeStatic } from "./serve-static.ts";
export { createServer, attachRouting } from "./server.ts";
export { Documentation } from "./documentation.ts";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors.ts";
export { testEndpoint, testMiddleware } from "./testing.ts";
export { Integration } from "./integration.ts";
export { EventStreamFactory } from "./sse.ts";

export { ez } from "./proprietary-schemas.ts";

// Convenience types
export type { Depicter } from "./documentation-helpers.ts";
export type { Producer } from "./zts-helpers.ts";

// Interfaces exposed for augmentation
export type { LoggerOverrides } from "./logger-helpers.ts";
export type { TagOverrides } from "./common-helpers.ts";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
import type {} from "qs"; // fixes TS2742 for attachRouting
export type { Routing } from "./routing.ts";
export type { FlatObject } from "./common-helpers.ts";
export type { Method } from "./method.ts";
export type { IOSchema } from "./io-schema.ts";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type.ts";
export type { ApiResponse } from "./api-response.ts";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  HeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security.ts";
