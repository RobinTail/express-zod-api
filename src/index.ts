import "./zod-plugin";

export { createConfig } from "./config-type";
export { AbstractEndpoint } from "./endpoint";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory";
export {
  getExamples,
  getMessageFromError,
  getStatusCodeFromError,
} from "./common-helpers";
export { createLogger } from "./logger";
export { createMiddleware } from "./middleware";
export {
  createResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler";
export { DependsOnMethod } from "./depends-on-method";
export { ServeStatic } from "./serve-static";
export { createServer, attachRouting } from "./server";
export { Documentation } from "./documentation";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors";
export { testEndpoint } from "./testing";
export { Integration } from "./integration";

export { ez } from "./proprietary-schemas";

// Convenience types
export type { Depicter } from "./documentation-helpers";
export type { Producer } from "./zts-helpers";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
export type { MockOverrides } from "./testing";
export type { Routing } from "./routing";
export type { LoggerOverrides } from "./logger";
export type { FlatObject } from "./common-helpers";
export type { Method } from "./method";
export type { IOSchema } from "./io-schema";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type";
export type { MiddlewareDefinition } from "./middleware";
export type { ResultHandlerDefinition } from "./result-handler";
export type { ApiResponse } from "./api-response";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  CustomHeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security";
