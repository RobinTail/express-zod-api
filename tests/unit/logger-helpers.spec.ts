import { BuiltinLogger } from "../../src";
import { BuiltinLoggerConfig } from "../../src/builtin-logger";
import {
  AbstractLogger,
  formatDuration,
  isLoggerInstance,
  isSeverity,
  isHidden,
} from "../../src/logger-helpers";

describe("Logger helpers", () => {
  describe("isSeverity()", () => {
    test.each(["debug", "info", "warn", "error"])(
      "should recognize %s",
      (subject) => {
        expect(isSeverity(subject)).toBeTruthy();
      },
    );
    test.each(["something", "", 123, Symbol.dispose])(
      "should reject others %#",
      (subject) => {
        expect(isSeverity(subject)).toBeFalsy();
      },
    );
  });

  describe("isHidden", () => {
    test.each([
      ["debug", "debug", false],
      ["debug", "info", true],
      ["debug", "warn", true],
      ["debug", "error", true],
      ["info", "debug", false],
      ["info", "info", false],
      ["info", "warn", true],
      ["info", "error", true],
      ["warn", "debug", false],
      ["warn", "info", false],
      ["warn", "warn", false],
      ["warn", "error", true],
      ["error", "debug", false],
      ["error", "info", false],
      ["error", "warn", false],
      ["error", "error", false],
    ] as const)(
      "should compare %s to %s with %s result",
      (subject, gate, result) => {
        expect(isHidden(subject, gate)).toBe(result);
      },
    );
  });

  describe("isLoggerInstance()", () => {
    test.each<Partial<BuiltinLoggerConfig>>([
      { level: "silent" },
      { level: "debug", color: false },
      { level: "info", color: true },
      { level: "warn", depth: 5 },
      { level: "warn", depth: null },
      { level: "warn", depth: Infinity },
    ])("should invalidate built-in logger config %#", (sample) => {
      expect(isLoggerInstance(sample)).toBeFalsy();
    });

    test.each<AbstractLogger>([
      // issue #1605: should not allow methods
      { level: "debug", debug: () => {} },
      { level: "warn", error: () => {} },
      // issue #1772: similar to #1605, but the methods are in prototype
      new (class {
        level = "debug";
        debug() {}
      })(),
      Object.setPrototypeOf({ level: "debug" }, { debug: () => {} }),
      new BuiltinLogger({ level: "debug" }),
    ])("should validate logger instances %#", (sample) => {
      expect(isLoggerInstance(sample)).toBeTruthy();
    });
  });

  describe("formatDuration()", () => {
    test.each([
      1e-9, 1e-8, 1e-7, 1e-6, 1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1, 1e1, 1e2, 1e3,
      15e2, 1e4, 1e5, 1e6, 1e7, 1e8, 1e9,
    ])("%# should format %s ms", (duration) =>
      expect(formatDuration(duration)).toMatchSnapshot(),
    );
  });
});
