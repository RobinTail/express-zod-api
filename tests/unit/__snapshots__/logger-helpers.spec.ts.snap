// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Logger helpers > formatDuration() > 0 should format 1e-9 ms 1`] = `"1 picosecond"`;

exports[`Logger helpers > formatDuration() > 1 should format 0.000001 ms 1`] = `"1 nanosecond"`;

exports[`Logger helpers > formatDuration() > 2 should format 0.001 ms 1`] = `"1 microsecond"`;

exports[`Logger helpers > formatDuration() > 3 should format 1 ms 1`] = `"1 millisecond"`;

exports[`Logger helpers > formatDuration() > 4 should format 1000 ms 1`] = `"1 second"`;

exports[`Logger helpers > formatDuration() > 5 should format 1000000 ms 1`] = `"17 minutes"`;

exports[`Logger helpers > formatDuration() > 6 should format 1000000000 ms 1`] = `"16667 minutes"`;
